File: python/change.py
import argparse
from sys import argv
from argparse import ArgumentParser
from datetime import datetime
def get_max_val(dict_to_search:dict) -> str:
    """
    Returns the key of the max value of a dict
    """
    return max(dict_to_search,key = dict_to_search.get,default = 0) 

def print_map(dict_to_print:dict) -> None:
    d = {k:v for k,v in dict_to_print.items() if v != 0}
    for k,v in d.items():
        print(f"\t{k}: {v}")
    

def change(amt:float,print_to_screen = True):
    """
    This will make the optimal amount of change given the amount. The amount is represented in terms of dollars
    """
    mapping = {
        'Penny': 1,
        'Nickel': 5,
        'Dime': 10,
        'Quarter': 25,
        'Dollar1': 100,
        'Dollar5': 500,
        'Dollar10': 1000,
        'Dollar20': 2000,
        'Dollar50': 5000,
        'Dollar100': 10000
    }
    cent_val = amt * 1 ## convert the amount to cents to match the mapping
    change_amt = 0 
    change_map = {x:0 for x in mapping.keys()}

    while change_amt < cent_val:
        delta_amt = cent_val - change_amt ## the remaining balance of change that is due
        filtered = {k:v for k,v in mapping.items() if v <= delta_amt}
        max_key = get_max_val(filtered) ## the key of the max
        change_amt += mapping[max_key]
        change_map[max_key] += 1
    
    if print_to_screen:
        print_map(change_map)
        print("====================================")
    return change_amt

def read_args(file:str) -> list:
    # ret_list = []
    with open(file,'r') as f:
        line = f.read().split('\n')
    ret_list = list(map(lambda x: round(float(x),2),line)) 

    return list(map(lambda x: round(x * 100,0),ret_list))

def write_output(args:list) -> None:
    begin_time = datetime.now()
    for arg in args:
        change(arg,False)
    end_time = datetime.now()

    print(f"Python took {end_time - begin_time} to run")
    
        

if __name__  == '__main__':
    parser = ArgumentParser()
    parser.add_argument('-c','--compare',
    help= 'Will read from the args.txt file instead of reading arguments from the command line',
    action = 'store_true')
    args = parser.parse_args()
    if args.compare:
        text_args = read_args('args.txt')
        write_output(text_args)
    else:
        for (i,arg) in enumerate(argv):
            if i == 0:
                continue
            amt = float(arg) ## convert command line args to float
            print(f"For ${amt}:")
            amt = round(amt * 100,0) ## convert the value to cents, and round just in case of floating point errors 
            change(amt)

File: julia/change.jl
using ArgParse
## Overload the Base max function to find the pair with the max value
function Base.max(d::Base.Dict{String,Int64})
	ret_val = 0 ## will be updated
	ret_key = nothing ## will be updated
	for (k,v) in d
		if v > ret_val
			ret_val = v
			ret_key = k
		end
	end
	return ret_key => ret_val
end

function filter_dict(d::Base.Dict{String,Int64},amt::Base.Number)
	temp = filter(x->x.second <= amt,d)
	return temp
end

function print_dict(coins::Base.Array{String,1},d::Base.Dict{String,Int64})
	for c in coins
		v = get(d,c,0) ## retrieve the value from the map
		if v != 0
			println("\t$c: $v")
		end
	end
end

function make_change(amt::Base.Integer)
	mapping = Dict(
		"Dollar100" => 10000,
		"Dollar50" => 5000,
		"Dollar20" => 2000,
		"Dollar10" => 1000,
		"Dollar5" => 500,
		"Dollar1" => 100,
		"Quarter" => 25,
		"Dime" => 10,
		"Nickel" => 5,
		"Penny" =>1
	)
	change_map = Dict(k => 0 for k in keys(mapping))
	change_amt = 0 
	while change_amt < amt
		delta = amt - change_amt
		temp_d = filter_dict(mapping,delta)
		ret_max = max(temp_d) ## use our customized max function
		max_k = ret_max.first ## the key
		max_v = ret_max.second ## the value
		change_amt += max_v ## update the total change
		delta += max_v ## update the difference between the tallied change and the requested amount
		change_map[max_k] += 1 ## increment the change map 
	end
	return filter(x->x.second > 0,change_map)

end

function parse_commandline()
	s = ArgParseSettings()

	@add_arg_table s begin 
		"-c"
			help = "Will read arguments from a file instead of the command line"
			action = :store_true
	end
	return parse_args(s)
end


function main(amt::String)
	coins = ["Penny","Nickel","Dime","Quarter","Dollar1","Dollar5","Dollar10","Dollar20","Dollar50","Dollar100"]
	amt_val = parse(Float32,amt)
	println("For \$$amt:")
	cent_amt = Integer(round(amt_val * 100,digits = 0))
	filtered_dict = make_change(cent_amt)
	coins_filt = filter(c -> c in keys(filtered_dict),coins)
	print_dict(coins_filt,filtered_dict)
	println("====================================")
end

## Add additional method to the main function
function main()
	coins = ["Penny","Nickel","Dime","Quarter","Dollar1","Dollar5","Dollar10","Dollar20","Dollar50","Dollar100"]
	file_data = ""
	file_data = read("args.txt",String)
	file_split = split(file_data,'\n')
	for args in file_split 
		if args == ""
			continue
		end
		amt_val = parse(Float32,args)
		cent_amt = Integer(round(amt_val * 100,digits = 0))
		filtered_dict = make_change(cent_amt)
		coins_filt = filter(c -> c in keys(filtered_dict),coins)
		# print_dict(coins_filt,filtered_dict)
		# println("====================================")
	end
end

## Run the script 
if length(parse_commandline()) > 0
	@time main()
else
	for a in ARGS 
		main(a)
	end
end
